### ISSUE TITLE
Doesn't start

### ISSUE THREAD

**Issue by cchantep** [deadcore/reactive-mongo-channel-not-found-error#1]
At `./start.sh` : 

```
Starting test_rs0-1_1 ... 
Starting test_rs0-0_1 ... 
Starting test_rs0-1_1
Starting test_rs0-0_1
Starting test_rs0-2_1 ... 
Starting test_rs0-0_1 ... done
Starting test_toxiproxy_1 ... 
Starting test_toxiproxy_1 ... done
Starting test_toxiproxy-configurer_1 ... 
Starting test_toxiproxy-configurer_1 ... done
Attaching to test_rs0-2_1, test_rs0-1_1, test_rs0-0_1, test_toxiproxy_1, test_toxiproxy-configurer_1
rs0-1_1                 | 2017-08-25T16:24:25.828+0000 I CONTROL  [initandlisten] MongoDB starting : pid=1 port=27028 dbpath=/data/db 64-bit host=rs0-1
rs0-1_1                 | 2017-08-25T16:24:25.828+0000 I CONTROL  [initandlisten] db version v3.4.7
rs0-1_1                 | 2017-08-25T16:24:25.828+0000 I CONTROL  [initandlisten] git version: cf38c1b8a0a8dca4a11737581beafef4fe120bcd
rs0-2_1                 | 2017-08-25T16:24:25.483+0000 I CONTROL  [initandlisten] MongoDB starting : pid=1 port=27029 dbpath=/data/db 64-bit host=rs0-2
rs0-0_1                 | 2017-08-25T16:24:26.742+0000 I CONTROL  [initandlisten] MongoDB starting : pid=1 port=27027 dbpath=/data/db 64-bit host=rs0-0
rs0-2_1                 | 2017-08-25T16:24:25.483+0000 I CONTROL  [initandlisten] db version v3.4.7
rs0-1_1                 | 2017-08-25T16:24:25.828+0000 I CONTROL  [initandlisten] OpenSSL version: OpenSSL 1.0.1t  3 May 2016
rs0-0_1                 | 2017-08-25T16:24:26.742+0000 I CONTROL  [initandlisten] db version v3.4.7
rs0-2_1                 | 2017-08-25T16:24:25.483+0000 I CONTROL  [initandlisten] git version: cf38c1b8a0a8dca4a11737581beafef4fe120bcd
rs0-1_1                 | 2017-08-25T16:24:25.828+0000 I CONTROL  [initandlisten] allocator: tcmalloc
rs0-0_1                 | 2017-08-25T16:24:26.742+0000 I CONTROL  [initandlisten] git version: cf38c1b8a0a8dca4a11737581beafef4fe120bcd
rs0-2_1                 | 2017-08-25T16:24:25.483+0000 I CONTROL  [initandlisten] OpenSSL version: OpenSSL 1.0.1t  3 May 2016
toxiproxy_1             | time="2017-08-25T16:24:27Z" level="info" msg="API HTTP server starting" host="0.0.0.0" port="8474" version="2.1.0" 
rs0-1_1                 | 2017-08-25T16:24:25.828+0000 I CONTROL  [initandlisten] modules: none
rs0-0_1                 | 2017-08-25T16:24:26.742+0000 I CONTROL  [initandlisten] OpenSSL version: OpenSSL 1.0.1t  3 May 2016
rs0-2_1                 | 2017-08-25T16:24:25.483+0000 I CONTROL  [initandlisten] allocator: tcmalloc
toxiproxy_1             | time="2017-08-25T16:24:28Z" level="info" msg="Started proxy" name="rs0-1" proxy="172.18.0.5:27018" upstream="rs0-1:27028" 
rs0-1_1                 | 2017-08-25T16:24:25.828+0000 I CONTROL  [initandlisten] build environment:
rs0-0_1                 | 2017-08-25T16:24:26.742+0000 I CONTROL  [initandlisten] allocator: tcmalloc
rs0-2_1                 | 2017-08-25T16:24:25.483+0000 I CONTROL  [initandlisten] modules: none
rs0-1_1                 | 2017-08-25T16:24:25.828+0000 I CONTROL  [initandlisten]     distmod: debian81
rs0-0_1                 | 2017-08-25T16:24:26.742+0000 I CONTROL  [initandlisten] modules: none
rs0-2_1                 | 2017-08-25T16:24:25.483+0000 I CONTROL  [initandlisten] build environment:
rs0-1_1                 | 2017-08-25T16:24:25.828+0000 I CONTROL  [initandlisten]     distarch: x86_64
toxiproxy_1             | time="2017-08-25T16:24:28Z" level="info" msg="Started proxy" name="rs0-0" proxy="172.18.0.5:27017" upstream="rs0-0:27027" 
rs0-0_1                 | 2017-08-25T16:24:26.742+0000 I CONTROL  [initandlisten] build environment:
rs0-2_1                 | 2017-08-25T16:24:25.483+0000 I CONTROL  [initandlisten]     distmod: debian81
rs0-1_1                 | 2017-08-25T16:24:25.828+0000 I CONTROL  [initandlisten]     target_arch: x86_64
rs0-0_1                 | 2017-08-25T16:24:26.742+0000 I CONTROL  [initandlisten]     distmod: debian81
rs0-2_1                 | 2017-08-25T16:24:25.483+0000 I CONTROL  [initandlisten]     distarch: x86_64
rs0-1_1                 | 2017-08-25T16:24:25.828+0000 I CONTROL  [initandlisten] options: { net: { port: 27028 }, replication: { oplogSizeMB: 128, replSet: "rs0" }, security: { authorization: "enabled", keyFile: "/opt/mongodb-keyfile" }, storage: { mmapv1: { smallFiles: true } } }
rs0-0_1                 | 2017-08-25T16:24:26.742+0000 I CONTROL  [initandlisten]     distarch: x86_64
rs0-2_1                 | 2017-08-25T16:24:25.483+0000 I CONTROL  [initandlisten]     target_arch: x86_64
rs0-0_1                 | 2017-08-25T16:24:26.742+0000 I CONTROL  [initandlisten]     target_arch: x86_64
toxiproxy_1             | time="2017-08-25T16:24:28Z" level="info" msg="Started proxy" name="rs0-2" proxy="172.18.0.5:27019" upstream="rs0-2:27029" 
rs0-2_1                 | 2017-08-25T16:24:25.483+0000 I CONTROL  [initandlisten] options: { net: { port: 27029 }, replication: { oplogSizeMB: 128, replSet: "rs0" }, security: { authorization: "enabled", keyFile: "/opt/mongodb-keyfile" }, storage: { mmapv1: { smallFiles: true } } }
rs0-1_1                 | 2017-08-25T16:24:25.832+0000 I -        [initandlisten] Detected data files in /data/db created by the 'wiredTiger' storage engine, so setting the active storage engine to 'wiredTiger'.
rs0-0_1                 | 2017-08-25T16:24:26.742+0000 I CONTROL  [initandlisten] options: { net: { port: 27027 }, replication: { oplogSizeMB: 128, replSet: "rs0" }, security: { authorization: "enabled", keyFile: "/opt/mongodb-keyfile" }, storage: { mmapv1: { smallFiles: true } } }
rs0-1_1                 | 2017-08-25T16:24:25.832+0000 I STORAGE  [initandlisten] 
rs0-2_1                 | 2017-08-25T16:24:25.486+0000 I -        [initandlisten] Detected data files in /data/db created by the 'wiredTiger' storage engine, so setting the active storage engine to 'wiredTiger'.
rs0-1_1                 | 2017-08-25T16:24:25.832+0000 I STORAGE  [initandlisten] ** WARNING: Using the XFS filesystem is strongly recommended with the WiredTiger storage engine
rs0-2_1                 | 2017-08-25T16:24:25.486+0000 I STORAGE  [initandlisten] 
rs0-1_1                 | 2017-08-25T16:24:25.832+0000 I STORAGE  [initandlisten] **          See http://dochub.mongodb.org/core/prodnotes-filesystem
rs0-0_1                 | 2017-08-25T16:24:26.745+0000 I -        [initandlisten] Detected data files in /data/db created by the 'wiredTiger' storage engine, so setting the active storage engine to 'wiredTiger'.
rs0-2_1                 | 2017-08-25T16:24:25.486+0000 I STORAGE  [initandlisten] ** WARNING: Using the XFS filesystem is strongly recommended with the WiredTiger storage engine
rs0-1_1                 | 2017-08-25T16:24:25.832+0000 I STORAGE  [initandlisten] wiredtiger_open config: create,cache_size=2466M,session_max=20000,eviction=(threads_min=4,threads_max=4),config_base=false,statistics=(fast),log=(enabled=true,archive=true,path=journal,compressor=snappy),file_manager=(close_idle_time=100000),checkpoint=(wait=60,log_size=2GB),statistics_log=(wait=0),
rs0-2_1                 | 2017-08-25T16:24:25.486+0000 I STORAGE  [initandlisten] **          See http://dochub.mongodb.org/core/prodnotes-filesystem
rs0-0_1                 | 2017-08-25T16:24:26.745+0000 I STORAGE  [initandlisten] 
rs0-0_1                 | 2017-08-25T16:24:26.746+0000 I STORAGE  [initandlisten] ** WARNING: Using the XFS filesystem is strongly recommended with the WiredTiger storage engine
rs0-0_1                 | 2017-08-25T16:24:26.746+0000 I STORAGE  [initandlisten] **          See http://dochub.mongodb.org/core/prodnotes-filesystem
rs0-2_1                 | 2017-08-25T16:24:25.486+0000 I STORAGE  [initandlisten] wiredtiger_open config: create,cache_size=2466M,session_max=20000,eviction=(threads_min=4,threads_max=4),config_base=false,statistics=(fast),log=(enabled=true,archive=true,path=journal,compressor=snappy),file_manager=(close_idle_time=100000),checkpoint=(wait=60,log_size=2GB),statistics_log=(wait=0),
rs0-1_1                 | 2017-08-25T16:24:26.404+0000 W STORAGE  [initandlisten] Detected configuration for non-active storage engine mmapv1 when current storage engine is wiredTiger
rs0-0_1                 | 2017-08-25T16:24:26.746+0000 I STORAGE  [initandlisten] wiredtiger_open config: create,cache_size=2466M,session_max=20000,eviction=(threads_min=4,threads_max=4),config_base=false,statistics=(fast),log=(enabled=true,archive=true,path=journal,compressor=snappy),file_manager=(close_idle_time=100000),checkpoint=(wait=60,log_size=2GB),statistics_log=(wait=0),
rs0-0_1                 | 2017-08-25T16:24:27.474+0000 W STORAGE  [initandlisten] Detected configuration for non-active storage engine mmapv1 when current storage engine is wiredTiger
rs0-1_1                 | 2017-08-25T16:24:26.406+0000 I FTDC     [initandlisten] Initializing full-time diagnostic data capture with directory '/data/db/diagnostic.data'
rs0-2_1                 | 2017-08-25T16:24:26.034+0000 W STORAGE  [initandlisten] Detected configuration for non-active storage engine mmapv1 when current storage engine is wiredTiger
rs0-1_1                 | 2017-08-25T16:24:26.407+0000 I REPL     [initandlisten] Did not find local voted for document at startup.
rs0-0_1                 | 2017-08-25T16:24:27.476+0000 I FTDC     [initandlisten] Initializing full-time diagnostic data capture with directory '/data/db/diagnostic.data'
rs0-1_1                 | 2017-08-25T16:24:26.407+0000 I REPL     [initandlisten] Did not find local replica set configuration document at startup;  NoMatchingDocument: Did not find replica set configuration document in local.system.replset
rs0-2_1                 | 2017-08-25T16:24:26.036+0000 I FTDC     [initandlisten] Initializing full-time diagnostic data capture with directory '/data/db/diagnostic.data'
rs0-0_1                 | 2017-08-25T16:24:27.476+0000 I REPL     [initandlisten] Did not find local voted for document at startup.
rs0-2_1                 | 2017-08-25T16:24:26.037+0000 I REPL     [initandlisten] Did not find local voted for document at startup.
rs0-0_1                 | 2017-08-25T16:24:27.476+0000 I REPL     [initandlisten] Did not find local replica set configuration document at startup;  NoMatchingDocument: Did not find replica set configuration document in local.system.replset
rs0-1_1                 | 2017-08-25T16:24:26.407+0000 I NETWORK  [thread1] waiting for connections on port 27028
rs0-2_1                 | 2017-08-25T16:24:26.037+0000 I REPL     [initandlisten] Did not find local replica set configuration document at startup;  NoMatchingDocument: Did not find replica set configuration document in local.system.replset
rs0-0_1                 | 2017-08-25T16:24:27.476+0000 I NETWORK  [thread1] waiting for connections on port 27027
rs0-2_1                 | 2017-08-25T16:24:26.037+0000 I NETWORK  [thread1] waiting for connections on port 27029
toxiproxy-configurer_1  | {"name":"rs0-1","listen":"172.18.0.5:27018","upstream":"rs0-1:27028","enabled":true,"toxics":[]}{"name":"rs0-0","listen":"172.18.0.5:27017","upstream":"rs0-0:27027","enabled":true,"toxics":[]}{"name":"rs0-2","listen":"172.18.0.5:27019","upstream":"rs0-2:27029","enabled":true,"toxics":[]}test_toxiproxy-configurer_1 exited with code 0
```

> BTW `docker-compose -p test up` would be sufficient to start up a compose (even in a fresh env).

**Comment by deadcore** [comment]
Thanks for the `docker-compose` hint 

In saying that it should be fine, each mongo instance was started expecting a replica set. The replica set is configured in `./setup.sh`. It was to get around the problems associated with using `--auth` flag

Unless I am missing something?

**Comment by cchantep** [comment]
Nope, that's not working and the setup doesn't find expected containers.

**Comment by deadcore** [comment]
Ok one sec thats on me

**Comment by deadcore** [comment]
@cchantep Just forsed the docker name in `docker-compose` to `reactivemongoissue_rs0-00`. If you pull it should work now all being well

**Comment by cchantep** [comment]
Still same error : 

```
reactivemongoissue_rs0-02 | 2017-08-25T17:16:10.030+0000 I NETWORK  [thread1] waiting for connections on port 27029
reactivemongoissue_rs0-00 | 2017-08-25T17:16:10.141+0000 I NETWORK  [thread1] waiting for connections on port 27027
toxiproxy-configurer_1  | {"name":"rs0-1","listen":"172.18.0.5:27018","upstream":"rs0-1:27028","enabled":true,"toxics":[]}{"name":"rs0-0","listen":"172.18.0.5:27017","upstream":"rs0-0:27027","enabled":true,"toxics":[]}{"name":"rs0-2","listen":"172.18.0.5:27019","upstream":"rs0-2:27029","enabled":true,"toxics":[]}test_toxiproxy-configurer_1 exited with code 0
```

**Comment by deadcore** [comment]
@cchantep Thats fine, just then run `setup.sh` and `Bootstrap` in a seperate console

**Comment by cchantep** [comment]
Still same error :

```
toxiproxy-configurer_1  | {"name":"rs0-1","listen":"172.18.0.5:27018","upstream":"rs0-1:27028","enabled":true,"toxics":[]}{"name":"rs0-0","listen":"172.18.0.5:27017","upstream":"rs0-0:27027","enabled":true,"toxics":[]}{"name":"rs0-2","listen":"172.18.0.5:27019","upstream":"rs0-2:27029","enabled":true,"toxics":[]}test_toxiproxy-configurer_1 exited with code 0
```

**Comment by cchantep** [comment]
Still same error:

```
toxiproxy-configurer_1  | {"name":"rs0-1","listen":"172.18.0.5:27018","upstream":"rs0-1:27028","enabled":true,"toxics":[]}{"name":"rs0-0","listen":"172.18.0.5:27017","upstream":"rs0-0:27027","enabled":true,"toxics":[]}{"name":"rs0-2","listen":"172.18.0.5:27019","upstream":"rs0-2:27029","enabled":true,"toxics":[]}test_toxiproxy-configurer_1 exited with code 0
```

**Comment by deadcore** [comment]
@cchantep That would be expected. The docker instances should spin up, then the configurer's will run through and inject the proxies into Toxiproxy and keep everything open.

Once that is done a separate terminal is needed to run `./setup.sh` to configure the replica set and create the users to mimic our environment

Finally the error happens when you run `Bootstrap.scala` with SBT using the command `sbt "run-main Bootstrap"`

An example of this would be 
![screenshot from 2017-08-29 14-41-17](https://user-images.githubusercontent.com/1455141/29824294-0a1c1ecc-8cc9-11e7-82d8-24a87369b096.png)

**Comment by cchantep** [comment]
Would rather declare the container dependencies in the compose and run the `setup.sh` when the `reactivemongoissue_rs0-00` container is executed

**Comment by deadcore** [comment]
@cchantep I think this should be enough to demonstrate the problem

**Comment by cchantep** [comment]
(a) What I can see with logging is that, with simulated latency, sometimes the primary is set unavailable due to all channels being closed (as Netty detects none respond timely according the connection options).

(b) Then the channels are established again (periodic `ConnectAll`/`RefreshAll`), the primary responds to a new IsMaster request, but as authentication is required it is not yet back online.

The auth request is sent to the primary, but sometimes, according latency, a new timeout happens, resulting in all the primary channels being closed again, and so the primary again unavailable (requiring another isMaster+auth).

(c) After that, the primary eventually completes the step isMaster + auth, to be set available.

**Comment by deadcore** [comment]
@cchantep Would you be able to attach the logs  from `Bootstrap` please. A couple of us have tried this and it always raises the `ChannelNotFoundException`

**Comment by cchantep** [comment]
Yes this raise a `ChannelNotFoundException` if an operation is executed after (a) (all channel closed), even if (b) the primary is detected (has responded to the isMaster probe) but has not yet responded to the subsequent authentication request.

If the time required by the primary to responds both and successively isMaster and auth exceed the connection option and/or the failover strategy, then it's legit this result in an exception.

**Comment by deadcore** [comment]
@cchantep,
I believe we are experiencing different errors. Whilst during the latency injection does cause `ChannelNotFoundException`, the one we are experiencing is when the pool has been drained, new connections have been made but the authentication has failed. Meaning the connections are ready but end up with this exception:
`ChannelNotFound: MongoError['Channel not found from the primary node: 'localhost:27017' { authenticated:0, connected:10, channels:10 } (Supervisor-2/Connection-3)']`

Once the latency stops and the mongo system has recovered (as what happens in Mongo Atlas), no write request succeeds and the whole of the reactive mongo system fails the above exception. We believe the problem is here https://github.com/ReactiveMongo/ReactiveMongo/blob/7c387ce64c03cd4576cb7548f85bb768c9e5bfff/driver/src/main/scala/core/actors/MongoDBSystem.scala#L1047


**This wouldn't be a problem if the system recovered however it requires our application to be restarted as it can not make any of the write requests**

**Comment by cchantep** [comment]
The exception, which corresponds to what I see in logging, means that's channel are back up (after timeout due to latency), not that authentication has not yet happen.

At some point, even with such latency both isMaster and auth are successful, w/o restarting the app (but with such latency such issue will happen often).

**Comment by deadcore** [comment]
@cchantep,

In which case are you able to run and provide the logs please?
 
The part where you mention
> At some point, even with such latency both isMaster and auth are successful, w/o restarting the app (but with such latency such issue will happen often). 

Does not stand true because ever after several hours, a healthy Mongo replica set which we can connect to externally our application is still throwing `ChannelNotFound: MongoError['Channel not found from the primary node: 'localhost:27017' { authenticated:0, connected:10, channels:10 } (Supervisor-2/Connection-3)']` exceptions on every write

**Comment by cchantep** [comment]
Enable the debug level and have a look at the logs. At first, "healthy Mongo replica set" won't have such latency.

**Comment by deadcore** [comment]
@cchantep Sorry I don't understand what you are asking for?

**Comment by cchantep** [comment]
For me the docker test case doesn't reproduce the same issue as the one reported on the [MailingList](https://groups.google.com/forum/?fromgroups#!topic/reactivemongo/crcsB9AdobM) about Atlas.

Both issues are reported as `ChannelNotFound`, but the one in the docker test case is due to simulated latency, and the driver can eventually get it up, even if of course such latency cannot be considered as a "healthy Mongo replica set".

The other `ChannelNotFound` is also indicating that no active channel can be found (not that no channel exists for the resolved primary), but as the driver is never able to resolve that, I guess the cause is different.

> The message will be updated to be more explicit.

The message indicates that there are 10 connected channels out of 10, but 0 authenticated.

So, the authentication settings, specially with Atlas (moreover for the Free Tier), need to be check:

```scala
import scala.concurrent.ExecutionContext.Implicits.global
import reactivemongo.api._

val driver = MongoDriver()
val con = driver.connection(...).get

con.authenticate(dbName, dbUser, dbPass)
```

> [RM-SBT-Playground](http://github.com/cchantep/RM-SBT-Playground) can be used to test connection settings.

```xml
<configuration debug="false">
  <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
    <!-- encoders are assigned the type
         ch.qos.logback.classic.encoder.PatternLayoutEncoder by default -->
    <encoder>
      <pattern>%d{HH:mm:ss.SSS} %-5level %logger{10} - %msg [%thread]%n</pattern>
    </encoder>
  </appender>

  <logger name="reactivemongo" level="info" />
  <logger name="reactivemongo.core" level="debug" />
  
  <root level="DEBUG">
    <appender-ref ref="STDOUT" />
  </root>
</configuration>
```

**Comment by deadcore** [comment]
@cchantep 

Apologies for the late reply, no laptop on the whole weekend :tada:.

I am struggling to see how changing the way you authenticate at the start helps in this case?

Unless I am mistaken the issues on the mailing list and here are the same. Coming from the original Chka mentions this:
```
Channel not found from the nodes: 
'node1' [Secondary] { authenticated:10, connected:10, channels:10 }; 
'node2' [Secondary] { authenticated:10, connected:10, channels:10 }; 
'node3' [Primary] { authenticated:0, connected:10, channels:10 } (Supervisor-5/Connection-6); 
[(1500905953501,
ConnectAll$IsMaster(830, {{
NodeSet None 
Node[node1: Secondary (10/10 available connections), latency=2], auth=Set(Authenticated(admin,my-db)) | 
Node[node2: Secondary (10/10 available connections), latency=2], auth=Set(Authenticated(admin,my-db)) | 
Node[node3: Primary (10/10 available connections), latency=1], auth=Set(Authenticated(admin,my-db)) }})), ...] 
```
Which has the characteristic `'ChannelNotFound { authenticated:0, connected:10, channels:10 }`. Once that has happened Chka got "hundreds of this error logged per min".


From now we are really struggling to understand what you are asking for? We have proven that the issue happens, so far there has been no resolution

**Comment by cchantep** [comment]
What's is proven/seen is that's the primary is not authenticated, for whatever reason. As suggested, make sure the credentials are ok using `.authenticate`, also enable the logging as suggested in the troubleshooting doc to investigate the env.

**Comment by deadcore** [comment]
@cchantep 
We did that when we initially looked at this long before raising the issue. We thought it may be network related as it was isolated to one availability zone but this was pure fluke.

Unfortunately everything pointed at Reactive Mongo (hence why we have raised the issue) and thus why we have invested the time to organise this.

The credentials authenticate fine because we looked at that when we built up the first iterations of this issue
